### ISSUE TITLE
Enhancement: [no-invalid-void-type] Allow `void` in return type type instantiations

### ISSUE THREAD

**Issue by jun-sheaf** [typescript-eslint/typescript-eslint#8755]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-invalid-void-type

### Description

`void`s used in return type type instantiations are almost always (I say almost, but it's probably always) correct. E.g. `S | void` or `Result<void, E>`. A deeper explanation can be found here: https://github.com/typescript-eslint/typescript-eslint/issues/5752#issuecomment-2024114126

### Fail

```TypeScript
N/A
```

### Pass

```TypeScript
(state: U) => U | void;

type Result<T, E = Error> = {type: 'success', value: T} | {type: 'error', value: E};
(state: U) => Result<void>;
```

**Comment by jun-sheaf** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/5752

**Comment by Josh-Cena** [comment]
I'm going to call this a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/8113 as we work out the areas we need to rework on.
